### PR TITLE
[`FanControl`] Refactor to use `AttributeAccessInterface` and `CommandHandlerInterface` alongside application delegate

### LINKED ISSUE #27261: Re-Architect Fan Control Server Code and Add Fan Control Delegate
### Description

From chats with various engineers, we discussed how the overall design goal was to:

1. Each cluster instance (i.e. endpoint, cluster pair) is represented by an object.
2. That object has some built-in behavior and some ways for the application to customize its behavior as needed.
3. There is no weird global function that then has to do dispatch going on.

The Fan Control Server currently relies on the global attribute callback function for the implementing device to get attribute changes so we need to improve this by offering some kind of Fan Delegate class description that implementers can use when instantiating. 

The `Step` command was also introduced for 1.2 and its effect is entirely application specific so we need a way to offload this to the application, and the above delegate idea would be perfect for this.

### Task List

- [ ] Re-architect Fan Control server to use `CommandHandlerInterface` and `AttributeAccessInterface`
- [ ] Design a virtual FanDelegate to handle writes and commands
- [ ] Add to all-clusters-app

### PR DESCRIPTION
**Pull request by mhazley**
### Description

Fan Control required a delegate to handle the application-specific Step Command that was introduced as part of the Fal Release Spec Update. 

I wasn't a massive fan of the Global Attribute Changed Callback that would have to be used to monitor for updates to the Fan Control cluster, so given that we had some extra weeks before TE2, I updated this to use the `AttributeAccessInterface` and the `CommandHandlerInterface` so that FanControl becomes an instantiable object alongside a `Delegate` class that allows for callbacks to cover all write and command scenarios. 

Fixes #27261

### CODE REVIEW

File: src/app/clusters/fan-control-server/fan-control.h [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+#include <app/CommandHandlerInterface.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace FanControl {
+
+class Delegate;
+
+class Instance : public AttributeAccessInterface, public CommandHandlerInterface
```
Comment (bzbarsky-apple) [inline src/app/clusters/fan-control-server/fan-control.h:32]: Why does this need to use CommandHandlerInterface?

File: src/darwin/Framework/CHIP/templates/availability.yaml
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/availability.yaml:1]: These changes are wrong and would break backwards compat on Darwin.  Please do not make these changes at all; I will deal with the (much more complicated) changes to this file that are needed to handle a rename like this.  Especially because we have already renamed the enum type, but not in the same release as the values... which is going to make this pretty painful all around.

This is the part that is an automatic "changes requested".

File: src/app/zap-templates/zcl/data-model/chip/fan-control-cluster.xml [hunk 2]
```diff
@@ -44,8 +44,8 @@ limitations under the License.
     <item name="Off/Low/High" value="0x01"/>
     <item name="Off/Low/Med/High/Auto" value="0x02"/>
     <item name="Off/Low/High/Auto" value="0x03"/>
-    <item name="Off/On/Auto" value="0x04"/>
-    <item name="Off/On" value="0x05"/>
+    <item name="Off/High/Auto" value="0x04"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/fan-control-cluster.xml:47]: The XML change should probably be in a separate PR from the implementation change, so it can be carefully reviewed in terms of its impact on SDK APIs.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
I have not looked at the actual implementation details yet, pending an answer the the CommandHandlerInterface question.

File: src/darwin/Framework/CHIP/templates/availability.yaml
Comment (mhazley) [reply src/darwin/Framework/CHIP/templates/availability.yaml:1]: Thanks for the clarification, I will revert this.

File: src/app/zap-templates/zcl/data-model/chip/fan-control-cluster.xml (hunk 2 above)
Comment (mhazley) [reply src/app/zap-templates/zcl/data-model/chip/fan-control-cluster.xml:47]: Noted, will revert for this PR.

File: src/app/clusters/fan-control-server/fan-control.h (hunk 1 above)
Comment (mhazley) [reply src/app/clusters/fan-control-server/fan-control.h:32]: It does not _have_ to use the `CommandHandlerInterface`. 

I have a simpler implementation on a branch that uses the `emberAfFanControlClusterStepCallback` to a very small delegate that handles the command. 

I (perhaps wrongly) perceived that we wanted to move away from that style of implementation to a more object-oriented approach where we instantiate an object per cluster and have an application delegate linked to it. 

So, for that reason, I used `CommandHandlerInterface` so it was easier to offload the command to the correct delegate.

I did this work because we had some extra time before TE2 and I (again, perhaps wrongly) perceived this as the way we wanted to move forward. 

Happy to discuss and also happy to revert to the previous approach if you would prefer, I went this direction because I thought it was desirable and FanControl had been provisional until now.

Comment (bzbarsky-apple) [reply src/app/clusters/fan-control-server/fan-control.h:32]: So here are the tradeoffs: using CommandHandlerInterface in its current state forces all the code for an optional command to be included, even for a device that does not implement that command.  But it does allow the  architecture you describe, which may be what we want to move toward.

Using the static cluster callback to handle the command means the unused code can get stripped out at link time if the command is not actually supported, but then requires some sort of lookup table to find the right delegate, or requires having a single delegate across all endpoints or something.

It might be worth just documenting why we are using CommandHandlerInterface here.

File: src/app/clusters/fan-control-server/fan-control.h [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+     * @param newMode The new FanMode value.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR HandleFanModeChange(FanModeEnum newMode) = 0;
+
+    /**
+     * This function will be called when the FanMode attribute is changed to Off. The server will change the SpeedSetting and
+     * PercentSetting in this case so are included in the callback.
+     * @param newPercentSetting The new PercentSetting value.
+     * @param newSpeedSetting The new SpeedSetting value (can be ignored if the feature is not supported)
```
Comment (bzbarsky-apple) [inline src/app/clusters/fan-control-server/fan-control.h:255]:
```suggestion
     * @param newSpeedSetting The new SpeedSetting value.  If the MultiSpeed feature is not supported, will not have a clearly defined value and must be ignored.
```

File: src/app/clusters/fan-control-server/fan-control.h [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     *
+     */
+    virtual CHIP_ERROR HandleFanModeOff(DataModel::Nullable<Percent> newPercentSetting,
+                                        DataModel::Nullable<uint8_t> newSpeedSetting) = 0;
+
+    /**
+     * This function will be called when the FanMode attribute is changed to Auto. The server will change the SpeedSetting and
```
Comment (bzbarsky-apple) [inline src/app/clusters/fan-control-server/fan-control.h:263]: Similar comments as for the Off case.

File: src/app/clusters/fan-control-server/fan-control.h [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+    virtual CHIP_ERROR HandleFanModeAuto(DataModel::Nullable<Percent> newPercentSetting,
+                                         DataModel::Nullable<uint8_t> newSpeedSetting)
+    {
+        return CHIP_ERROR_NOT_IMPLEMENTED;
+    }
+
+    /**
+     * This function will be called when the PercentSetting or SpeedSetting attributes are changed.
+     * @param newPercentSetting The new PercentSetting value.
+     * @param newSpeedSetting The new SpeedSetting value (can be ignored if the feature is not supported)
```
Comment (bzbarsky-apple) [inline src/app/clusters/fan-control-server/fan-control.h:278]: See above for wording suggestion.

File: src/app/clusters/fan-control-server/fan-control.h [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR HandlePercentSpeedSettingChange(DataModel::Nullable<Percent> newPercentSetting,
+                                                       DataModel::Nullable<uint8_t> newSpeedSetting) = 0;
+
+    /**
+     * This function will be called when the RockSetting attribute is changed.
```
Comment (bzbarsky-apple) [inline src/app/clusters/fan-control-server/fan-control.h:285]: This, and the other callbacks, need to be very clear that they will be called when the relevant attributes are changed by _Matter interactions_, but not when the attributes are changed via calls to the cluster's public API, right?

It might be better to use "written to" instead of "changed" to make that clearer, but might also just need a longer description.

Also should document whether the values passed to these functions will have already been constraint-checked based on the spec constraints (yes, they will be if the comments above are fixed).  The delegate in all-clusters-app could then stop doing all sorts of checks it does right now.

File: src/app/clusters/fan-control-server/fan-control.h [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+    /**
+     * This function will be called when the AirflowDirection attribute is changed.
+     * @param newAirflowDirection The new AirflowDirection value.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR HandleAirflowDirectionChange(AirflowDirectionEnum newAirflowDirection) { return CHIP_ERROR_NOT_IMPLEMENTED; }
+
+    /**
+     * This function will be called when the Step Command is received and the STEP feature is enabled in the Feature Map
+     * @param stepDirection The direction of the step command.
```
Comment (bzbarsky-apple) [inline src/app/clusters/fan-control-server/fan-control.h:307]: Again, should document that this has been constraint-checked.

File: src/app/clusters/fan-control-server/fan-control.h [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+     * @param wrap Whether the step should wrap around to the other end of the sequence.
+     * @param lowestOff Whether the lowest setting should be treated as off.
+     * @return Returns an interation model status code indicating the result of the command.
+     */
+    virtual chip::Protocols::InteractionModel::Status HandleStep(StepDirectionEnum stepDirection, chip::Optional<bool> pWrap,
```
Comment (bzbarsky-apple) [inline src/app/clusters/fan-control-server/fan-control.h:312]: Why the extra chip prefixes all over here?

File: src/app/tests/suites/TestFanControl.yaml [hunk 9]
```diff
@@ -40,18 +40,6 @@ tests:
       response:
           value: 3
 
-    - label: "Write fan mode sequence"
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/TestFanControl.yaml:43]: Why was this removed?

File: src/app/tests/suites/TestFanControl.yaml [hunk 10]
```diff
@@ -68,19 +56,21 @@ tests:
       command: "readAttribute"
       attribute: "SpeedSetting"
       response:
-          value: 84
+          value: 9
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/TestFanControl.yaml:59]: Why the change?  And to all the other values in this file?

File: src/app/tests/suites/TestFanControl.yaml [hunk 11]
```diff
@@ -68,19 +56,21 @@ tests:
...
 
     - label: "Write percent setting"
       command: "writeAttribute"
       attribute: "PercentSetting"
       arguments:
           value: null
+      response:
+          error: WRITE_IGNORED
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/TestFanControl.yaml:73]: As discussed above, this is wrong.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Review by bzbarsky-apple — COMMENTED** [review]
And to be clear: I am _very_ worried about this scale of rewrite at this late date, especially given our clear lack of test coverage.  Are these changes blocking fall release?  If not, it might be better to make them early in the next cycle instead of near the end of this one, and in the meantime improve the testing situation...

**Comment by mhazley** [comment]
>  I am very worried about this scale of rewrite at this late date, especially given our clear lack of test coverage. Are these changes blocking fall release?

I am happy to not do this PR - I thought I was trying to be proactive but I take your point completely - these changes are not blocking fall release and I can revert to the simpler delegate to support the STEP command only, thus leaving the rest of the code basically unchanged.

Thanks for the thorough review regardless.
